torch_xla/csrc/runtime/pjrt_registry.cc

```diff
@@ -92,29 +115,27 @@ InitializePjRt(const std::string& device_type) {
     int global_process_rank = sys_util::GetEnvInt("RANK", local_process_rank);
     int local_world_size = sys_util::GetEnvInt("LOCAL_WORLD_SIZE", 1);
     int global_world_size = sys_util::GetEnvInt("WORLD_SIZE", local_world_size);
-    std::string master_addr =
-        runtime::sys_util::GetEnvString("MASTER_ADDR", "localhost");
-    std::string port = runtime::sys_util::GetEnvString(
-        "XLA_COORDINATOR_PORT", XlaCoordinator::kDefaultCoordinatorPort);
-
-    bool spmd = sys_util::GetEnvBool("XLA_USE_SPMD", false);
-    std::optional<std::set<int>> allowed_devices;
-    if (!spmd) {
-      allowed_devices = std::set{local_process_rank};
-    }
 
     std::shared_ptr<xla::KeyValueStoreInterface> kv_store;
-    if (global_world_size > 1) {
-      // Use the distributed key-value store from DistributedRuntimeClient.
-      coordinator = std::make_unique<XlaCoordinator>(
-          global_process_rank, global_world_size, master_addr, port);
-      std::shared_ptr<xla::DistributedRuntimeClient> distributed_client =
-          coordinator->GetClient();
-      kv_store = xla::GetDistributedKeyValueStore(distributed_client,
-                                                  /*key_prefix=*/"gpu:");
+    std::optional<std::set<int>> allowed_devices;
+    bool spmd = sys_util::GetEnvBool("XLA_USE_SPMD", false);
```

Comment:
Conditioning on SPMD mode here could cause issues using `xr.use_spmd()` after the runtime has been initialized.

Is it correct to say that `allowed_devices` is only needed in the MP case? If so, can we invert the logic to check for MP using one of the env vars instead of checking for SPMD mode?

Reply:
Is there ever a reason to call `xr.use_spmd()` after the runtime is initialized? In any case, I think we can also assume that if `LOCAL_WORLD_SIZE=1`, then we can use all of the devices (which should be compatible with SPMD)

Reply:
AFAIK there's not a strong use case at the moment, but for example our unit tests will check `xr.global_runtime_device_count()` before calling `xr.use_spmd()`. Keeping the runtime independent of SPMD mode was something we wanted to maintain, cc @yeounoh

Reply:
That's a good point!

> Conditioning on SPMD mode here could cause issues using xr.use_spmd() after the runtime has been initialized.

This seems to be a downside of using `xr.use_spmd()` as opposed to a env flag `XLA_USE_SPMD=1`. With the latter, it's less flexible but less error-prone. It guarantee we'll use spmd mode at the beginning. With the former, it may also impact other SPMD special cases: user does something pytorch ops, then call `xr.use_spmd()`, then continue to do something else.

> can we invert the logic to check for MP using one of the env vars instead of checking for SPMD mode?
> I think we can also assume that if LOCAL_WORLD_SIZE=1, then we can use all of the devices

I thought about this route but didn't feel it's clean. What if the user has 2 GPU machines and she wants to use 1 GPU device on each machine and to do multi-host training.

Reply:
> I think we can also assume that if LOCAL_WORLD_SIZE=1, then we can use all of the devices

Perhaps we also need to check GLOBAL_WORLD_SIZE:
```
if LOCAL_WORLD_SIZE==1:
  if GLOBAL_WORLD_SIZE>1: # multi-host-single-process
    initialize coordinator service
    allowd_devices={current_device}
  else: # single-host-single-process
    do nothing
else: multi-process for single-host and multi-host
  initialize coordinator service
  allowd_devices={current_device}
```